### DISCUSSION TITLE
Should we use UDP or TCP for network communication, and what’s the difference in the context of a game like this???

### DISCUSSION THREAD

**Discussion by kherldhussein** [kh3rld/prisoners-dilemma#43] [Q&A] [answered]
Choosing between UDP (User Datagram Protocol) and TCP (Transmission Control Protocol) can impact the performance and reliability of network communication in the game.

**Comment by kh3rld** [comment, marked as answer]
I recommend sticking with TCP. IT ensures that all data packets arrive in order and without errors, which is important for turn-based games where every move needs to be processed in the correct sequence. UDP, on the other hand, is faster but doesn’t guarantee delivery or order, making it more suitable for real-time games where speed is critical, and some data loss is acceptable. If you’re interested, you can check out more about these protocols on [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Glossary/TCP).

**Reply by kherldhussein** [reply]
Would using WebSockets provide any advantages over TCP for our PD game? If so, how should we implement it???

**Reply by kh3rld** [reply]
WebSockets actually work on top of TCP, so you get all the benefits of TCP (like reliable and ordered delivery of messages) but with some added advantages. WebSockets are ideal for applications that require real-time, bidirectional communication. For a turn-based game like ours, using WebSockets can simplify the process of maintaining a persistent connection between the client and server, which can be more efficient than opening and closing TCP connections for each action or turn.

Implementing WebSockets would allow us to send messages to the server whenever a player makes a move and receive updates immediately, without the overhead of HTTP headers. This makes the game feel more responsive and keeps both players in sync more smoothly.

To implement WebSockets in Go, you can use libraries like [Gorilla WebSocket](https://github.com/gorilla/websocket). It’s a well-maintained library that provides a straightforward API for upgrading HTTP connections to WebSockets and handling communication. You can find a great tutorial on how to set up WebSockets in Go [here](https://youtu.be/pKpKv9MKN-E?t=1).

WebSockets could be a great fit if you want to make the game communication more efficient and maintainable, especially as you add more real-time features or expand multiplayer capabilities.
